### PR TITLE
Updated summary with pod's parent information

### LINKED ISSUE #731: Summary data doesn't show information about the pod's parent except for deployment
### **Problem Statement** 

The latest Discovery Engine generates summary information for a pod, stores it in a database, and pushes the same information to the publisher. The table structure has a field `deployment_name` which will be populated when the pod is the child of a deployment. This information will be empty if the pod is deployed using a StatefulSet/DaemonSet/ReplicaSet.

**Current Schema**

```
CREATE TABLE `system_summary` ( 
  `id` INTEGER AUTO_INCREMENT, 
  `cluster_name` varchar(50) DEFAULT NULL, 
  `cluster_id` int DEFAULT NULL,  
  `workspace_id` int DEFAULT NULL, 
  `namespace_name` varchar(50) DEFAULT NULL, 
  `namespace_id` int DEFAULT NULL, 
  `container_name` varchar(50) DEFAULT NULL, 
  `container_image` varchar(100) DEFAULT NULL, 
  `container_id` varchar(150) DEFAULT NULL, 
  `podname` varchar(50) DEFAULT NULL, 
  `operation` varchar(10) DEFAULT NULL,   
  `labels` varchar(100) DEFAULT NULL, 
  `deployment_name` varchar(50) DEFAULT NULL, 
  `source` varchar(100) DEFAULT NULL, 
  `destination` varchar(100) DEFAULT NULL, 
  `destination_namespace` varchar(50) DEFAULT NULL, 
  `destination_labels` varchar(50) DEFAULT NULL,  
  `type` varchar(10) DEFAULT NULL, 
  `ip` int DEFAULT NULL,  
  `port` varchar(10) DEFAULT NULL, 
  `protocol` varchar(10) DEFAULT NULL, 
  `bindport` varchar(10) DEFAULT NULL, 
  `bindaddr` varchar(10) DEFAULT NULL, 
  `action` varchar(10) DEFAULT NULL, 
  `count` int NOT NULL,   
  `updated_time` bigint NOT NULL, 
  PRIMARY KEY (`id`)  
);
```

**Proposed Change**
To make sure that the pod's parent information is updated correctly on the database we would need the following
1. Change the `deployment_name` to `parent_name` to make sure that pods deployed using other resources are also getting updated on the database
2. Add a new field `parent_type` to store the pod's parent type (StatefulSet/ReplicaSet/DaemonSet/Deployments etc)

**Proposed Schema**

```
CREATE TABLE `system_summary` ( 
  `id` INTEGER AUTO_INCREMENT, 
  `cluster_name` varchar(50) DEFAULT NULL, 
  `cluster_id` int DEFAULT NULL,  
  `workspace_id` int DEFAULT NULL, 
  `namespace_name` varchar(50) DEFAULT NULL, 
  `namespace_id` int DEFAULT NULL, 
  `container_name` varchar(50) DEFAULT NULL, 
  `container_image` varchar(100) DEFAULT NULL, 
  `container_id` varchar(150) DEFAULT NULL, 
  `podname` varchar(50) DEFAULT NULL, 
  `operation` varchar(10) DEFAULT NULL,   
  `labels` varchar(100) DEFAULT NULL, 
  `parent_name` varchar(50) DEFAULT NULL, 
  `parent_type` varchar(50) DEFAULT NULL, 
  `source` varchar(100) DEFAULT NULL, 
  `destination` varchar(100) DEFAULT NULL, 
  `destination_namespace` varchar(50) DEFAULT NULL, 
  `destination_labels` varchar(50) DEFAULT NULL,  
  `type` varchar(10) DEFAULT NULL, 
  `ip` int DEFAULT NULL,  
  `port` varchar(10) DEFAULT NULL, 
  `protocol` varchar(10) DEFAULT NULL, 
  `bindport` varchar(10) DEFAULT NULL, 
  `bindaddr` varchar(10) DEFAULT NULL, 
  `action` varchar(10) DEFAULT NULL, 
  `count` int NOT NULL,   
  `updated_time` bigint NOT NULL, 
  PRIMARY KEY (`id`)  
);
```

### PR DESCRIPTION
**Pull request by vishnusomank**
Fixes: https://github.com/accuknox/discovery-engine/issues/731

Related PRs --> https://github.com/kubearmor/kubearmor-client/pull/299

This PR contains the following changes:

- Updated summary information to include pod's parent_name and parent_type
- Updated functions to retrieve daemonset details
- Updated constants with deployment, replicaset, statefulset and daemonset values
- Updated protobuf definitions to include parent_name and parent_type to pod information

Please refer [299#issuecomment-1561132322](https://github.com/kubearmor/kubearmor-client/pull/299#issuecomment-1561132322) for the updated output

### CODE REVIEW

File: src/protobuf/v1/observability/observability.proto [hunk 1]
```diff
@@ -36,6 +37,7 @@ message Response{
     repeated CiliumSummData IngressData = 11;
     repeated CiliumSummData EgressData = 12;
     repeated SysNwSummaryData BindConnection = 13;
+    string DeployType = 14;
```
Comment (rajaSahil) [inline src/protobuf/v1/observability/observability.proto]: What does DeployType signify over here?

Comment (rajaSahil) [reply src/protobuf/v1/observability/observability.proto]: Shouldn't this be ResourceType?

File: src/protobuf/v1/observability/observability.proto [hunk 2]
```diff
@@ -20,6 +20,7 @@ message Request{
     string Type = 6;
     bool Aggregate = 7;
     string DeployName = 8;
```
Comment (rajaSahil) [inline src/protobuf/v1/observability/observability.proto]: Can we name this as ResourceName?

File: src/types/constants.go [hunk 3]
```diff
@@ -48,4 +48,10 @@ const (
 
 	// max no. of tries to connect to kubearmor-relay
 	Maxtries = 6
+
+	// Resource types
+	K8sDeploymentType = "Deployment"
+	K8sStatefulSetsType = "StatefulSet"
+	K8sReplicaSetsType = "ReplicaSet"
+	K8sDaemonSetsType = "DaemonSet"
```
Comment (rajaSahil) [inline src/types/constants.go]: Do we also need to add for jobs?

File: src/protobuf/v1/observability/observability.proto (hunk 1 above)
Comment (nyrahul) [reply src/protobuf/v1/observability/observability.proto]: k8s uses the term resource owner for this.

**Comment by vishnusomank** [comment]
@rajaSahil I've updated the code based on your comment. Can you please review again?
cc: @nyrahul

File: src/cluster/k8sClientHandler.go [hunk 4]
```diff
@@ -557,6 +555,97 @@ func GetStatefulSetsFromK8sClient() []types.Deployment {
...
+		}
+	}
+	return results
+}
+
+// ================= //
+// == DaemonSets == //
+// ================= //
+
+func GetDaemonSetsFromK8sClient() []types.Deployment {
```
Comment (Ankurk99) [inline src/cluster/k8sClientHandler.go]: Can we create a single function and then pass the k8s objects as argument? like GetObjectSetsFromK8sClient(object), then pass daemonset/deployment etc. as a value to it. It will greatly reduce the code size

Comment (vishnusomank) [reply src/cluster/k8sClientHandler.go]: Addressed @Ankurk99

File: src/cluster/k8sClientHandler.go [hunk 5]
```diff
@@ -489,81 +444,96 @@ func GetReplicaSetsFromK8sClient() []types.Deployment {
...
+			// get ReplicaSet list from k8s api client
+			replicaSets, err := client.AppsV1().ReplicaSets("").List(context.Background(), metav1.ListOptions{})
+			if err != nil {
+				log.Error().Msg(err.Error())
+				return results
 			}
+			for _, rs := range replicaSets.Items {
+				results = append(results, GenerateResourceList(rs.OwnerReferences,rs.Name,rs.Namespace,types.K8sReplicaSetsType,rs.Spec.Selector.MatchLabels)...)
+			}
+		case types.K8sJobsType:
```
Comment (rajaSahil) [inline src/cluster/k8sClientHandler.go:492]: Do we also need to consider cronjobs? Although it creates new jobs periodically.

File: src/conf/local-file.yaml [hunk 6]
```diff
@@ -119,6 +119,7 @@ recommend:
   operation-mode: 1                       # 1: cronjob | 2: one-time-job
   cron-job-time-interval: "1h0m00s"       # format: XhYmZs
   recommend-host-policy: true
+  admission-controller-policy: false
```
Comment (rajaSahil) [inline src/conf/local-file.yaml:122]: Why this change?

File: src/libs/dbHandler.go [hunk 7]
```diff
@@ -433,7 +433,7 @@ func upsertSysSummarySQL(db *sql.DB, summary types.SystemSummary, timeCount type
 	summary.Labels = strings.Join(sortedLabels, ",")
 
 	queryString := `cluster_name = ? and cluster_id = ? and workspace_id = ? and namespace_name = ? and namespace_id = ? and container_name = ? and container_image = ? 
-					and podname = ? and operation = ? and labels = ? and deployment_name = ? and source = ? and destination = ? 
+					and podname = ? and operation = ? and labels = ? and parent_name = ? and parent_type = ? and source = ? and destination = ?
```
Comment (rajaSahil) [inline src/libs/dbHandler.go]: Can we name table columns as resource_type and resource_name?

File: src/libs/sqliteHandler.go [hunk 8]
```diff
@@ -722,7 +722,8 @@ func CreateSystemSummaryTableSQLite(cfg types.ConfigDB) error {
 			"	`podname` varchar(50) DEFAULT NULL," +
 			"	`operation` varchar(10) DEFAULT NULL," +
 			"	`labels` varchar(100) DEFAULT NULL," +
-			"	`deployment_name` varchar(50) DEFAULT NULL," +
+			"	`parent_name` varchar(50) DEFAULT NULL," +
+			"	`parent_type` varchar(50) DEFAULT NULL," +
```
Comment (rajaSahil) [inline src/libs/sqliteHandler.go]: resource_type and resource_name ?

File: src/libs/sqliteHandler.go [hunk 9]
```diff
@@ -1604,17 +1609,17 @@ func GetPodNamesSQLite(cfg types.ConfigDB, filter types.ObsPodDetail) ([]string,
...
 	db := connectSQLite(cfg, config.GetCfgObservabilityDBName())
 	defer db.Close()
 
-	resDeployNames := []string{}
+	resDeployNames := map[string]string{}
 
 	var results *sql.Rows
 	var err error
 
 	// Get podnames from system table
```
Comment (rajaSahil) [inline src/libs/sqliteHandler.go]:
```suggestion
	// Get resource type and resource names from system table.
```

File: src/libs/sqliteHandler.go [hunk 10]
```diff
@@ -1644,13 +1653,14 @@ func GetDeployNamesSQLite(cfg types.ConfigDB, filter types.ObsPodDetail) ([]stri
...
-		var locDeployName string
+		var locDeployName, locDeployType string
 		if err := results.Scan(
 			&locDeployName,
+			&locDeployType,
 		); err != nil {
 			return nil, err
 		}
-		resDeployNames = append(resDeployNames, locDeployName)
+		resDeployNames[locDeployName] = locDeployType
```
Comment (rajaSahil) [inline src/libs/sqliteHandler.go:1663]: There can be possibility of having same resource names in different namespace, so when getting all the entries from table, are we handling this case?

File: src/observability/summarizer.go [hunk 11]
```diff
@@ -126,6 +130,22 @@ func convertSysLogToSysSummaryMap(syslogs []*pb.Alert) {
...
+				if sysSummary.Deployment != "" {
+					continue
+				}
+				sysSummary.Deployment = d.Name
+				sysSummary.Workload.Type = d.Type
+				break
+			}
+		}
+		if sysSummary.Deployment == "" {
+			sysSummary.Workload.Type = "Pod"
```
Comment (rajaSahil) [inline src/observability/summarizer.go]: Is it correct to consider pod as a workload type?

**Comment by seswarrajan** [comment]
Currently putting this on hold as we are handling summary through different CLI tool
